### PR TITLE
[Improve](TabletSchemaCache) reduce duplicated memory consumption for column name and column path

### PR DESCRIPTION
**Pull request by eldenmoon**
… column name and column path

Both could be reference to related field in TabletColumn.And use shared_ptr for TabletColumn in TabletSchema for later memory reuse

## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

File: be/src/olap/in_list_predicate.h [hunk 1]
```diff
@@ -36,13 +36,6 @@
 #include "vec/common/string_ref.h"
 #include "vec/core/types.h"
 
-template <>
-struct std::equal_to<doris::StringRef> {
```
[the remark this answers is not in the document]
Comment (eldenmoon) [reply be/src/olap/in_list_predicate.h:40]: it's redundant and lead to compile error when using std::unordered_map<StringRef, ..>

File: be/src/olap/tablet_schema.h [hunk 2]
```diff
@@ -172,29 +181,30 @@ class TabletColumn {
     std::string _default_value;
 
     bool _is_decimal = false;
-    int32_t _precision;
-    int32_t _frac;
+    int32_t _precision = -1;
```
[the remark this answers is not in the document]
Comment (eldenmoon) [reply be/src/olap/tablet_schema.h:184]: init default value

File: be/src/olap/tablet_schema.cpp [hunk 3]
```diff
@@ -806,12 +809,11 @@ void TabletSchema::append_column(TabletColumn column, ColumnType col_type) {
     }
     if (column.is_variant_type()) {
         ++_num_variant_columns;
-        if (column.path_info().empty()) {
+        if (!column.has_path_info()) {
             const std::string& col_name = column.name_lower_case();
             vectorized::PathInData path(col_name);
             column.set_path_info(path);
         }
-        _field_path_to_index[column.path_info()] = _num_columns;
```
[the remark this answers is not in the document]
Comment (eldenmoon) [reply be/src/olap/tablet_schema.cpp:814]: put to line 830 to ref path

File: be/src/olap/tablet_schema.cpp [hunk 4]
```diff
@@ -820,21 +822,21 @@ void TabletSchema::append_column(TabletColumn column, ColumnType col_type) {
     } else if (UNLIKELY(column.name() == VERSION_COL)) {
         _version_col_idx = _num_columns;
     }
+    _field_id_to_index[column.unique_id()] = _num_columns;
```
[the remark this answers is not in the document]
Comment (eldenmoon) [reply be/src/olap/tablet_schema.cpp:825]: using std::move

File: be/src/olap/tablet_schema.cpp [hunk 5]
```diff
@@ -910,9 +917,9 @@ void TabletSchema::init_from_pb(const TabletSchemaPB& schema, bool ignore_extrac
         if (column.is_variant_type()) {
             ++_num_variant_columns;
         }
-        _field_name_to_index[column.name()] = _num_columns;
-        _field_id_to_index[column.unique_id()] = _num_columns;
-        _cols.emplace_back(std::move(column));
+        _cols.emplace_back(std::make_shared<TabletColumn>(std::move(column)));
```
[the remark this answers is not in the document]
Comment (eldenmoon) [reply be/src/olap/tablet_schema.cpp:920]: yes

File: be/src/vec/common/schema_util.cpp [hunk 6]
```diff
@@ -689,7 +689,9 @@ void rebuild_schema_and_block(const TabletSchemaSPtr& original,
         vectorized::PathInDataBuilder full_root_path_builder;
         auto full_root_path =
                 full_root_path_builder.append(parent_column.name_lower_case(), false).build();
-        flush_schema->mutable_columns()[variant_pos].set_path_info(full_root_path);
+        TabletColumn new_col = flush_schema->column(variant_pos);
```
[the remark this answers is not in the document]
Comment (eldenmoon) [reply be/src/vec/common/schema_util.cpp:692]: shared_ptr is used `mutable_columns ` should be deleted

**Comment by eldenmoon** [comment]
run buildall

File: be/src/olap/tablet_schema.h [hunk 7]
```diff
@@ -279,8 +289,8 @@ class TabletSchema {
     Status have_column(const std::string& field_name) const;
     const TabletColumn& column_by_uid(int32_t col_unique_id) const;
     TabletColumn& mutable_column_by_uid(int32_t col_unique_id);
-    const std::vector<TabletColumn>& columns() const;
-    std::vector<TabletColumn>& mutable_columns();
+    void relace_column(size_t pos, TabletColumn new_col);
```
[the remark this answers is not in the document]
Comment (eldenmoon) [reply be/src/olap/tablet_schema.h]: done

**Comment by eldenmoon** [comment]
run buildall

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by qidaye — APPROVED** [review]

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Review by xiaokang — APPROVED** [review]
LGTM

**Pull request merged**